service/storage_service.cc

```diff
@@ -1166,9 +1166,34 @@ class topology_coordinator {
             auto reason = ::format(
                 "insert CDC generation data (UUID: {})", gen_uuid);
             co_await update_topology_state(std::move(guard), {std::move(mutation), builder.build()}, reason);
+            break;
         }
+        case global_topology_request::keyspace_rf_change: {
+            slogger.info("raft topology: new keyspace RF change requested");
+            auto tmptr = get_token_metadata_ptr();
+            auto ks_name = _topo_sm._topology.new_keyspace_rf_change_ks_name;
+            auto rf_per_dc = _topo_sm._topology.new_keyspace_rf_change_rf_per_dc;
+
+            locator::tablet_map tablets = tmptr->tablets().get_tablet_map(schema_ptr{}->id());
+            const auto& dcs = tmptr->get_topology().get_datacenters();
+            for (auto tablet : ks_tablets) {
+                tablet_id = tablet.id();
+                locator::tablet_replica_set replicas = tablets.get_tablet_info(tablet_id).replicas;
+                // TODO: should be based on network_topology_strategy::allocate_tablets_for_new_table, basically:
+                // When new RF < old RF, take the most loaded endpoints from load_sketch and drop extra redundant replicas from these endpoints
+                // Otherwise add new replicas from the least loaded endpoints
+                slogger.debug("Allocated tablets for {}.{} ({}): {}", s->ks_name(), s->cf_name(), s->id(), tablets);
+                tablet_transition_info transition{ .next = replicas };
+                tablets.set_tablet_transition_info(tablet_id, transition);
```

Comment:
You also need to modify tablet state machine to represent the new kind of migration. That means extending system.tablets with a new column "transition" which holds transition kind (rf_change | migration), and extending tablet_transition_info to represent it. And changing handle_tablet_migration() to jump from allow_write_both_read_old stage to end_migration directly instead of write_both_read_old if transition kind is rf_change.

Reply:
> I thought that in storage_service::alter_tablets_keyspace I should generate a `global_topology_request::keyspace_rf_change` request, attach RF/DC data to it and send it, so that when it's handled in `handle_global_request`, mutations are going to be generated. Can you help me understand it please?

Yes, in alter_tablets_keyspace you generate mutations which create the request, but in the global request handler you generate mutations which start tablet migrations. In any case, you always change group0 state by mutations, not by modifying in-memory structures like tablet_map directly. It's a global state which is managed by raft and to keep it globally consistent all updates must go through raft machinery.